openedx/openedx-platform
# Badging: Core Refactor

**Background**: This implements the a refactoring of Badging to be more flexible, creating a framework for developers to make different badges for different events and situations. Previously, a badge could only be awarded on course completion.
**Discussions/Specs**: https://docs.google.com/document/d/1ob-leRHV97agPGw5ys9uASXrcsU8qf51bzYlKH3mBEM/edit?usp=sharing
**Jira Ticket**: https://openedx.atlassian.net/browse/SOL-1362
**Partner Information**: 3rd party-hosted open edX instance, for an edX solutions client.
**Sandbox**: LMS at http://pr10556.sandbox.opencraft.com CMS at http://studio.pr10556.sandbox.opencraft.com

Deviations from spec:

As is usually the case, there are a few deviations from the spec that came up in implementation. Here are the most notable:
1. The get_badge_class function is on BadgeClass and not on a custom manager. It did not appear to require that much hardware.
2. A few extra fields have been added to BadgeClass and BadgeAssertion. BadgeClass now has 'mode' for backward compatibility, and a 'criteria' field, which is required by the OpenBadges spec. BadgeAssertions now have an 'assertion_url' field that points to the hosted JSON assertion.
3. SVG support is not here-- this wasn't in the spec but in the comments. As this implementation arrived before Badgr's support for SVG landed, it is not currently supported in the LMS.

Testing instructions:

This version of edx-platform should behave identically on the outside to how it has before. The demo course is currently set up on the sandbox to make it easy to pass. Just run through the exam section and answer those questions right. You will need to re-open the course by changing its close date so that it's in the future, enroll a user, answer the questions in the exam, then set the schedule to be closed, and run the certificate generation command as root:

```
sudo -u edxapp /edx/bin/python.edxapp /edx/app/edxapp/edx-platform/manage.py lms --settings aws ungenerated_certs -c course-v1:edX+DemoX+Demo_Course --insecure
```

This will generate a certificate for your user, which you should then be able to see in the dashboard. Click view for the certificate, and attempt to download the badge.
